### ISSUE TITLE
no-unused-vars doesn't recognize function arrays

### ISSUE THREAD

**Issue by a-gerescher** [eslint/eslint#12978]
**Tell us about your environment**

* **ESLint Version:** 6.8.0
* **Node Version:** 8.11.4
* **npm Version:** 6.9.0

**What parser (default, Babel-ESLint, etc.) are you using?** default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
"eslintConfig": {
    "env": {
      "browser": true,
      "es6": true
    },
    "extends": [
      "eslint:recommended",
      "eslint-config-developit"
    ],
    "globals": {
      "Atomics": "readonly",
      "SharedArrayBuffer": "readonly"
    },
    "parserOptions": {
      "ecmaVersion": 2018,
      "sourceType": "module",
      "ecmaFeatures": {
        "jsx": true
      }
    },
    "rules": {
      "indent": "off",
      "brace-style": [2, "1tbs", { "allowSingleLine": true }],
      "no-prototype-builtins": "off",
      "prefer-const": "off",
      "no-empty": [
        "warn",
        {
          "allowEmptyCatch": true
        }
      ]
    },
    "overrides": []
  }
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
const specialState = () => {
  let t = [ () => {} ];

  for (const x of t) {
    x();
  }
}
```

```bash

```

**What did you expect to happen?**

No warnings about unused vars.

**What actually happened? Please include the actual, raw output from ESLint.**

x is unused.

**Are you willing to submit a pull request to fix this bug?**

**Comment by mdjermanovic** [comment]
Hi @a-gerescher, thanks for the issue!

I can't reproduce this with default parser in [ESLint Online Demo](https://eslint.org/demo#eyJ0ZXh0IjoiY29uc3Qgc3BlY2lhbFN0YXRlID0gKCkgPT4ge1xuICBsZXQgdCA9IFsgKCkgPT4ge30gXTtcblxuICBmb3IgKGNvbnN0IHggb2YgdCkge1xuICAgIHgoKTtcbiAgfVxufSIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6OSwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6eyJjb25zdHJ1Y3Rvci1zdXBlciI6MiwiZm9yLWRpcmVjdGlvbiI6MiwiZ2V0dGVyLXJldHVybiI6Miwibm8tYXN5bmMtcHJvbWlzZS1leGVjdXRvciI6Miwibm8tY2FzZS1kZWNsYXJhdGlvbnMiOjIsIm5vLWNsYXNzLWFzc2lnbiI6Miwibm8tY29tcGFyZS1uZWctemVybyI6Miwibm8tY29uZC1hc3NpZ24iOjIsIm5vLWNvbnN0LWFzc2lnbiI6Miwibm8tY29uc3RhbnQtY29uZGl0aW9uIjoyLCJuby1jb250cm9sLXJlZ2V4IjoyLCJuby1kZWJ1Z2dlciI6Miwibm8tZGVsZXRlLXZhciI6Miwibm8tZHVwZS1hcmdzIjoyLCJuby1kdXBlLWNsYXNzLW1lbWJlcnMiOjIsIm5vLWR1cGUta2V5cyI6Miwibm8tZHVwbGljYXRlLWNhc2UiOjIsIm5vLWVtcHR5IjoyLCJuby1lbXB0eS1jaGFyYWN0ZXItY2xhc3MiOjIsIm5vLWVtcHR5LXBhdHRlcm4iOjIsIm5vLWV4LWFzc2lnbiI6Miwibm8tZXh0cmEtYm9vbGVhbi1jYXN0IjoyLCJuby1leHRyYS1zZW1pIjoyLCJuby1mYWxsdGhyb3VnaCI6Miwibm8tZnVuYy1hc3NpZ24iOjIsIm5vLWdsb2JhbC1hc3NpZ24iOjIsIm5vLWlubmVyLWRlY2xhcmF0aW9ucyI6Miwibm8taW52YWxpZC1yZWdleHAiOjIsIm5vLWlycmVndWxhci13aGl0ZXNwYWNlIjoyLCJuby1taXNsZWFkaW5nLWNoYXJhY3Rlci1jbGFzcyI6Miwibm8tbWl4ZWQtc3BhY2VzLWFuZC10YWJzIjoyLCJuby1uZXctc3ltYm9sIjoyLCJuby1vYmotY2FsbHMiOjIsIm5vLW9jdGFsIjoyLCJuby1wcm90b3R5cGUtYnVpbHRpbnMiOjIsIm5vLXJlZGVjbGFyZSI6Miwibm8tcmVnZXgtc3BhY2VzIjoyLCJuby1zZWxmLWFzc2lnbiI6Miwibm8tc2hhZG93LXJlc3RyaWN0ZWQtbmFtZXMiOjIsIm5vLXNwYXJzZS1hcnJheXMiOjIsIm5vLXRoaXMtYmVmb3JlLXN1cGVyIjoyLCJuby11bmRlZiI6Miwibm8tdW5leHBlY3RlZC1tdWx0aWxpbmUiOjIsIm5vLXVucmVhY2hhYmxlIjoyLCJuby11bnNhZmUtZmluYWxseSI6Miwibm8tdW5zYWZlLW5lZ2F0aW9uIjoyLCJuby11bnVzZWQtbGFiZWxzIjoyLCJuby11bnVzZWQtdmFycyI6Miwibm8tdXNlbGVzcy1jYXRjaCI6Miwibm8tdXNlbGVzcy1lc2NhcGUiOjIsIm5vLXdpdGgiOjIsInJlcXVpcmUteWllbGQiOjIsInVzZS1pc25hbiI6MiwidmFsaWQtdHlwZW9mIjoyfSwiZW52Ijp7fX19). There's a warning for `specialState`, but not for `x`.

I think you're actually using `babel-eslint` parser, which is set by `eslint-config-developit` in [this line](https://github.com/developit/eslint-config-developit/blob/684d34e298bdc12b7accd8a66c959f8d44ed0bd5/index.js#L2).

There was an issue with `no-unused-vars` and variables declared in `for-in` and `for-of` loops (babel/babel-eslint#791). This issue has been fixed and shouldn't appear anymore with `babel-eslint` version `v10.0.3` and above.

It seems that the actual published version of `eslint-config-developit` is still using `babel-eslint 8`.

Can you please try the following:

* Install the latest version of `babel-eslint`
* Add `"parser": "babel-eslint"` to `"eslintConfig"`

**Comment by a-gerescher** [comment]
Works as intended. Thank you very much.
